### PR TITLE
Combine precomputed values of filtered attribute sets

### LINKED ISSUE #3439: Align Async Counter behavior when used with filtered attributes.
### Problem Statement

If an attribute filter is used with an Async Counter the last value observed will be exported. This

```go
// When there is an Attribute Filter that only allows "foo" attributes.
 
ctr, _ := meter.AsyncInt64().Counter("testCounter")
...
ctr.Observe(ctx, 4, attribute.Int("Version", 1))
ctr.Observe(ctx, 5, attribute.Int("Version", 2))
```
Yields a Sum with a value of 5.

This has been discussed in the spec, and the expected output is a Sum with a value of 9.
https://github.com/open-telemetry/opentelemetry-specification/issues/2905

### Proposed Solution

Correct our instruments, filters, or aggregators to get the appropriate value.

### PR DESCRIPTION
**Pull request by MrAlias**
Fix #3439

When an attribute filter drops a distinguishing attribute during the aggregation of a precomputed sum that value needs to be added to any existing one for an equivalent set[^1]. This changes the current behavior of overriding the equivalent set's value.

[^1]: https://github.com/open-telemetry/opentelemetry-specification/issues/2905

### CODE REVIEW

File: sdk/metric/internal/sum.go
Comment (dashpole) [inline sdk/metric/internal/sum.go:273]: I thought precomputed sums were supposed to be reset each time they are collected?  Will this never be reset?

Comment (MrAlias) [reply sdk/metric/internal/sum.go:273]: For cumulative precomputed sums, it is the same as cumulative sums, they do not reset as the cumulative value is considered still relevant the next collection cycle. For delta cumulative sums, they do.

This change does not modify that behavior though. The cumulative sums preserve the previous value and the delta reset when `Aggregation` are called. The line you commented on is a shorthand for invoking this: https://github.com/open-telemetry/opentelemetry-go/blob/a724cf884287e04785eaa91513d26a6ef9699288/sdk/metric/internal/sum.go#L41-L45 which does not affect the collection reset functionality.

Comment (dashpole) [reply sdk/metric/internal/sum.go:273]: This is effectively replacing the line above, since we would previously call Aggregate from sdk/metric/internal/filter.go, right?:
```golang
func (s *precomputedSum[N]) Aggregate(value N, attr attribute.Set) {
	s.set(value, attr)
}
```

`set` has this implementation ([here](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/metric/internal/sum.go#LL35-L39C2): 

```golang
func (s *valueMap[N]) set(value N, attr attribute.Set) { // nolint: unused  // This is indeed used.
	s.Lock()
	s.values[attr] = value
	s.Unlock()
}
```

So I think it was resetting the precomputed one each time Aggregate was called?

Comment (MrAlias) [reply sdk/metric/internal/sum.go:273]: > So I think it was resetting the precomputed one each time Aggregate was called?

Correct, which is the specified behavior when the attribute set is not filtered, if a user reports a value multiple times for the same attribute set (which they shouldn't, but if they do) the last one needs to be used. When an attribute filter is used and it reduces a set to a previously recorded set, [that value needs to be added to the previous one](https://github.com/open-telemetry/opentelemetry-specification/issues/2905). Which is what is being done in these changes. Or, that is the intention. Do you not see that happening?

**Comment by MrAlias** [comment]
There is an edge case that this approach doesn't handle:

When the filtered observation is made prior to a non-filtered one, the non-filtered observation will override the previous filtered observation. Meaning ...

```
Observe(1, {"key", "value"}, {"user", "Alice"})
Observe(2, {"user", "Alice"})
```

when a filter of only `"user"` keys are allowed it will produce a value of `2` for the attribute set. Though the specification wants 3.

**Comment by MrAlias** [comment]
> There is an edge case that this approach doesn't handle:
> 
> When the filtered observation is made prior to a non-filtered one, the non-filtered observation will override the previous filtered observation. Meaning ...
> 
> ```
> Observe(1, {"key", "value"}, {"user", "Alice"})
> Observe(2, {"user", "Alice"})
> ```
> 
> when a filter of only `"user"` keys are allowed it will produce a value of `2` for the attribute set. Though the specification wants 3.

Handled in https://github.com/open-telemetry/opentelemetry-go/pull/3549/commits/b4630709647b58ab865dccf67202f628aa2b40ed which includes updated to tests to catch this.

File: sdk/metric/internal/sum.go
Comment (dashpole) [reply sdk/metric/internal/sum.go:273]: My reading of the old code was that it _always_ replaced the current value.  My reading of the new code is that it _always_ adds the new value to the current value.  If I have your test case:
```golang
tr, err := mtr.AsyncFloat64().Counter("afcounter")
if err != nil {
	return err
}
_, err = mtr.RegisterCallback([]instrument.Asynchronous{ctr}, func(ctx context.Context) {
	ctr.Observe(ctx, 1.0, attribute.String("foo", "bar"), attribute.Int("version", 1))
	ctr.Observe(ctx, 2.0, attribute.String("foo", "bar"))
	ctr.Observe(ctx, 1.0, attribute.String("foo", "bar"), attribute.Int("version", 2))
})
```

And your view:
```golang
WithView(NewView(
	Instrument{Name: "*"},
	Stream{AttributeFilter: func(kv attribute.KeyValue) bool {
		return kv.Key == attribute.Key("foo")
	}},
)),
```

The first time Collect is called, I would (correctly) get a metric with value 4 (1 + 2 + 1).

But the second time Collect is called, wouldn't I (incorrectly) get a metric with value 8 (where the correct value is still 4)?

Comment (MrAlias) [reply sdk/metric/internal/sum.go:273]: Gotcha, thanks for the breakdown. I think I see your point now.

I'm AFK currently, but will look into this tomorrow.

**Review by hanyuancheung — APPROVED** [review]

File: sdk/metric/internal/sum.go
Comment (MrAlias) [reply sdk/metric/internal/sum.go]: Updated, please take another look.

I think I have resolved this concern, and added tests to validate the solution. I'm going to resolve this, but if you think I missed something please unresolved.

File: sdk/metric/internal/sum_test.go
Comment (MadVikingGod) [inline sdk/metric/internal/sum_test.go:266]: Why is only the filtered dropped?  

Wouldn't this make the first measurement that was recorded have extra meaning?

Comment (MrAlias) [reply sdk/metric/internal/sum_test.go:266]: Please see this conversation: https://github.com/open-telemetry/opentelemetry-go/pull/3549#discussion_r1053619102

And to be clear, they are not dropped, they are reset given they were additions to a precomputed sum. The directly set sum, which the user is telling the SDK is the precomputed sum, has semantically different meaning.

**Comment by MrAlias** [comment]
For reviewers looking for more of an example: https://github.com/MrAlias/opentelemetry-go/pull/721

File: sdk/metric/internal/filter.go
Comment (dashpole) [inline sdk/metric/internal/filter.go:36]: why does agg being a filterAgg imply that it is precomputed?  I'm not entirely sure what a `filterAgg` is...

File: sdk/metric/internal/filter.go
Comment (dashpole) [inline sdk/metric/internal/filter.go:52]: this is only for non-precomputed filters, right?

File: sdk/metric/internal/filter.go
Comment (dashpole) [inline sdk/metric/internal/filter.go:81]: This comment confuses me slightly.  Why does it matter whether filtering is done?  No filter is equivalent to an empty filter, and just means a single element is aggregated.  I would think that a precomputed Aggregation would need to _always sum_, but would only end up summing once when no filtering is done.  The sum needs to be reset every aggregation "cycle".  Is this an optimization to prevent resetting and storing each time?

File: sdk/metric/internal/sum.go
Comment (dashpole) [inline sdk/metric/internal/sum.go:164]: am I correct in thinking that only one of [filtered, measured] will end up being set?  Is there a case where both are set?

**Review by dashpole — APPROVED** [review]
I still have trouble understanding the code structure a bit, but the tests help me believe that it works correctly.  Comments are just for my understanding, and possibly for additional comments in the code

File: sdk/metric/internal/sum.go
Comment (MrAlias) [reply sdk/metric/internal/sum.go:164]: If an observation is made that doesn't filter any attributes, `measured` is _set_. If an observation is made that does filter out attributes, `filtered` will be _added_ to. There may be a callback(s) that does both, so both may be used.

File: sdk/metric/internal/filter.go
Comment (MrAlias) [reply sdk/metric/internal/filter.go:81]: Precomputed aggregators don't add values, they set the measured value directly. The user is reporting the sum to record. They say the sun should be X and then in the next cycle they say the sun should be Y. If they added it would effectively saying in the second collection cycle the sum isn't Y but X+Y.

The problem is when one of these pre-computed sums has attributes removed and the resulting attributes collide with another pre-computed sum. In this case, the pre-computed some that had attributes removed needs to be added to the other.

File: sdk/metric/internal/filter.go
Comment (MrAlias) [reply sdk/metric/internal/filter.go:52]: Right. I'm AFK but can add a comment when I get back.

File: sdk/metric/internal/filter.go
Comment (MrAlias) [reply sdk/metric/internal/filter.go:36]: Yeah, could probably improve the name here. I'll take a look when I get back.

File: sdk/metric/internal/filter.go
Comment (MrAlias) [reply sdk/metric/internal/filter.go:57]: Added comment to docs.

File: sdk/metric/internal/filter.go
Comment (MrAlias) [reply sdk/metric/internal/filter.go]: Rename to `precomputedAggregator`. PTAL.

File: sdk/metric/internal/filter.go
Comment (dashpole) [reply sdk/metric/internal/filter.go]: To illustrate my point, this branch doesn't seem to impact the resulting behavior: https://github.com/open-telemetry/opentelemetry-go/commit/0fecb363a260565719f401667846e7a490725467

It still passes the complete tests in meter_test.go

Comment (dashpole) [reply sdk/metric/internal/filter.go]: The distinction between no filter and a noop filter means adding a filter that does nothing changes the resulting metrics, which seems wrong: https://github.com/open-telemetry/opentelemetry-go/commit/48dadf5e52f8a27917c48dc537cca6b8ef7d4c60

Comment (MrAlias) [reply sdk/metric/internal/filter.go]: > The distinction between no filter and a noop filter means adding a filter that does nothing changes the resulting metrics, which seems wrong: [48dadf5](https://github.com/open-telemetry/opentelemetry-go/commit/48dadf5e52f8a27917c48dc537cca6b8ef7d4c60)

I was not expecting that error in the comments, I expected the behavior to not change when introducing the no-op filter as well. I'm not able to reproduce the error though.

I pushed af451873e7aa903b86428e3d9b2b126ada4f572a to duplicate this in the upstream CI system to see if my issue with not producing the error is local.

Comment (MrAlias) [reply sdk/metric/internal/filter.go]: > To illustrate my point, this branch doesn't seem to impact the resulting behavior: [0fecb36](https://github.com/open-telemetry/opentelemetry-go/commit/0fecb363a260565719f401667846e7a490725467)
> 
> It still passes the complete tests in meter_test.go

I'm also not able to duplicate this change without the tests in `meter_test.go` failing.

```terminal 
--- FAIL: TestAsynchronousExample (0.00s)
    --- FAIL: TestAsynchronousExample/Cumulative (0.00s)
        meter_test.go:1207: [ScopeMetrics Metrics not equal:
            missing expected values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:60}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:53}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:5}}, Temporality:0x1, IsMonotonic:true}}
            unexpected additional values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:0}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:53}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:5}}, Temporality:0x1, IsMonotonic:true}}
            ]
    --- FAIL: TestAsynchronousExample/Delta (0.00s)
        meter_test.go:1282: [ScopeMetrics Metrics not equal:
            missing expected values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:0}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:6}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:5}}, Temporality:0x2, IsMonotonic:true}}
            unexpected additional values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:-60}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:6}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:5}}, Temporality:0x2, IsMonotonic:true}}
            ]
FAIL
FAIL	go.opentelemetry.io/otel/sdk/metric	0.025s
```

This is expected as a duplicate `Observer` of the same attribute set that is not filtered should be overwritten, not added (i.e. #3278 and #3350).

Comment (MrAlias) [reply sdk/metric/internal/filter.go]: > I pushed [af45187](https://github.com/open-telemetry/opentelemetry-go/commit/af451873e7aa903b86428e3d9b2b126ada4f572a) to duplicate this in the upstream CI system to see if my issue with not producing the error is local.

These changes passed upstream as well: https://github.com/open-telemetry/opentelemetry-go/actions/runs/3941799369/jobs/6744664098

Comment (MrAlias) [reply sdk/metric/internal/filter.go]: I'm wondering if our disconnect is in the base behavior here. If there is a `Int64ObservableCounter` for an SDK that filters its attributes to only allow the `user` attribute and the following observations are made:

- `obserer.Observe(10, {"user", "alice"})`
- `obserer.Observe(30, {"user", "alice"})`

The resulting measurement expected `30` for the sum datapoint with the `{"user", "alice"}` attributes, right?

Comment (dashpole) [reply sdk/metric/internal/filter.go]: I hadn't considered that you could record the same set of labels twice within an observation.  I'm used to that being disallowed in prom :).  But it would break this case.

Comment (dashpole) [reply sdk/metric/internal/filter.go]: I was wrong above as well, sorry.  I was running with both patches, rather than just the test commit.

Comment (dashpole) [reply sdk/metric/internal/filter.go]: Ok, so if I have the observations below, I get value = 30.
```
obserer.Observe(10)
obserer.Observe(30)
```

But then if I have the observations below, and a view that filters `key != foo`, I would get value = 40?

```
obserer.Observe(10, {"foo", "bar"})
obserer.Observe(30, {"foo", "bar"})
```

Fundamentally I want a filtered-out attribute to be the same as an attribute that never existed.  That implies that the fact that something was or was not filtered should never matter (in my ideal world).  Only the post-filter attribute set should matter.

Comment (MrAlias) [reply sdk/metric/internal/filter.go]: > Fundamentally I want a filtered-out attribute to be the same as an attribute that never existed. That implies that the fact that something was or was not filtered should never matter (in my ideal world). Only the post-filter attribute set should matter. So in theory, I would probably rather have the first example sum within a collection cycle.

So, to be clear, I agree with the approach you mention. However, this is not the behavior the broader OTel community expects as captured in https://github.com/open-telemetry/opentelemetry-specification/issues/2905 (and previously in https://github.com/open-telemetry/opentelemetry-specification/issues/1874).

The specification expects behavior laid out in [the supplemental guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/25f75bfde209630c60728039f27c041f70e5e738/specification/metrics/supplementary-guidelines.md#asynchronous-example-attribute-removal-in-a-view). Specifically, _additive_ for any precomputed sum that has had attributes removed.

I'm happy to switch the behavior if upstream the specification decides to do so, but with this PR I'm only trying to move our implementation into compliance with what the specification guidelines are.

Comment (dashpole) [reply sdk/metric/internal/filter.go]: https://github.com/open-telemetry/opentelemetry-specification/issues/3109

Comment (dashpole) [reply sdk/metric/internal/filter.go]: lets resolve this conversation and move forward with the fix (assuming other approvers are on-board)

File: sdk/metric/meter_test.go
Comment (dashpole) [inline sdk/metric/meter_test.go:1275]: Is this correct?  Seems to differ from the example in the supplementary guidelines: https://github.com/open-telemetry/opentelemetry-specification/blob/25f75bfde209630c60728039f27c041f70e5e738/specification/metrics/supplementary-guidelines.md#asynchronous-example-cumulative-temporality.  I'm trying to figure out where the spec says this

Comment (MrAlias) [reply sdk/metric/meter_test.go:1275]: I take the omission in the supplementary guidelines as not specifying what happens to the exported data for thread 1. For our SDK this goes back to the question of how to handle resets or dropped attribute sets, https://github.com/open-telemetry/opentelemetry-go/issues/3006. Until #3006 and the related https://github.com/open-telemetry/opentelemetry-specification/issues/1891#issuecomment-1194348335 are resolved, we don't have a way to persist state across collection cycles but also allow resets. And the delta implementation requires maintaining state across collections cycles to calculate the cumulative-to-delta conversion.

Comment (dashpole) [reply sdk/metric/meter_test.go:1275]: > And the delta implementation requires maintaining state across collections cycles to calculate the cumulative-to-delta conversion.

Does this imply the cumulative implementation also needs to maintain state in this case?  Haven't looked to see if this is possible implementation-wise.

Comment (MrAlias) [reply sdk/metric/meter_test.go:1275]: > Does this imply the cumulative implementation also needs to maintain state in this case? Haven't looked to see if this is possible implementation-wise.

It does, but only to ensure to false reset gaps are introduced.

I guess, given how the measured values are precomputed and set, you could expect the previous cycle's value to be reported again and wouldn't need to maintain state. But if we relied on this and the user failed to report the value for a cycle or two and then continued reporting it would introduce gaps (or zeros) in the reported cumulative metric value.

Comment (MrAlias) [reply sdk/metric/meter_test.go:1275]: > I guess, given how the measured values are precomputed and set, you could expect the previous cycle's value to be reported again and wouldn't need to maintain state. But if we relied on this and the user failed to report the value for a cycle or two and then continued reporting it would introduce gaps (or zeros) in the reported cumulative metric value.

For what its worth, this would also diverge from our other non-pre-computed cumulative sum and mean we had inconsistent output behavior for a `Counter` versus an `ObservableCounter`.

Comment (jmacd) [reply sdk/metric/meter_test.go:1275]: I think there's a potential for specification work on this question, but here's how I've been thinking about it. I feel that the reason for asynchronous instruments is to give the application a performance benefit when reporting metrics. Therefore, we should promote that benefit by not requiring the SDK to maintain additional state to ensure some kind of correctness and order.

It is the user's responsibility to behave, I think, which means asynchronous callbacks should not "forget" or duplicate attribute sets they report. The SDK has to go to some lengths to de-duplicate inputs and track forgotten attribute sets, and until convinced otherwise I think these should remain as unspecified behaviors.

Comment (dashpole) [reply sdk/metric/meter_test.go:1275]: > For what its worth, this would also diverge from our other non-pre-computed cumulative sum and mean we had inconsistent output behavior for a Counter versus an ObservableCounter.

I think this depends on how you define consistent user behavior with a Counter vs ObservableCounter.  We could say that incrementing a Counter is equivalent to an ObservableCounter that keeps returning 1 value higher than prior.  If, someday, the synchronous counter had a Delete(labels) function, the equivalent behavior in ObservableCounter would be to stop returning observations with those labels.  Today, we don't have an equivalent in the synchronous case, which I think is OK (prom, for example, allows removing label sets in the async case, but not in the sync case).

> It is the user's responsibility to behave, I think, which means asynchronous callbacks should not "forget" or duplicate attribute sets they report. The SDK has to go to some lengths to de-duplicate inputs and track forgotten attribute sets, and until convinced otherwise I think these should remain as unspecified behaviors.

+1.  To spell out my desired behavior further, IMO an async counter which stops reporting a label set (ignore errors for now) should not be reported in the cumulative or delta case, and any memory/caches associated with that label set should be immediately released.  The next observation of that label set should be reported as if it were completely new.  The start time should be the time Collect was previously invoked, since that was the last time we knew this series was absent.

> it would introduce gaps (or zeros) in the reported cumulative metric value.

This feels correct to me.  It gives the user exactly what they've asked for, which is no values during a period.

Comment (MrAlias) [reply sdk/metric/meter_test.go:1275]: I don't have strong feelings here.

@dashpole can you track this in a new issue? It doesn't seem like something that should block this PR given it is about default behavior and the PR is addressing filters, right?

Comment (dashpole) [reply sdk/metric/meter_test.go:1275]: This 100% shouldn't block this PR.  Sorry if that was not apparent

Comment (dashpole) [reply sdk/metric/meter_test.go:1275]: Left https://github.com/open-telemetry/opentelemetry-specification/issues/1891#issuecomment-1396167105

Comment (jmacd) [reply sdk/metric/meter_test.go:1275]: @dashpole I agree w/ your summary, thanks.

Comment (dashpole) [reply sdk/metric/meter_test.go:1504]: Here is an issue on our repo as well: https://github.com/open-telemetry/opentelemetry-go/issues/3605

**Comment by MrAlias** [comment]
Tracking link CI failure here: https://github.com/open-telemetry/opentelemetry-go/issues/2957#issuecomment-1398743407

**Pull request merged**
